### PR TITLE
UX Improvements

### PR DESCRIPTION
**Pull request by dimitrovmaksim**
Issues: 
closes https://github.com/LimeChain/matchstick/issues/119
closes https://github.com/LimeChain/matchstick/issues/310

Depends on: 
https://github.com/LimeChain/matchstick-as/pull/43

What it does:

- Adds `beforeAll` and `afterAll` functions
- Adds `beforeEach` and `afterEach` functions
- Adds `describe` with support for nested describes
- Fixes some issues with the logic for detecting if imported files have been changed.
- Now every `test.ts` file will be compiled into a separate test suite, even if multiple test files are grouped into a folder, e.g:
```sh
tests/
└── gravity
    ├── create.test.ts
    └── update.test.ts

1 directory, 2 files
```

will yield two test suites named `gravity/create` and `gravity/update`
- Updates running only the tests that were specified with a pattern
```sh
graph test gravity # this will run all test that contain the word gravity
```
1. [Basic Structure](https://github.com/LimeChain/matchstick/pull/273#user-content-basic-structure):
2. [How to test](https://github.com/LimeChain/matchstick/pull/273#user-content-how-to-test)

#### `describe(name: String , () => {})` - Defines an example/test group. 

**_Notes:_** 

- _Describes are not mandatory. You can still use test() outside of describe() blocks_

Example: 

```
import {describe, test } from "matchstick-as/assembly/index"

describe(“Example test group”, () => {
    test(“A test case”, () => {…})
	
    test(“Another test case”, () => {…})
})
```

Nested `describe()` example:

```
import {describe, test } from "matchstick-as/assembly/index"

describe(“Outside describe”, () => {
  describe("Inner describe", () => {
     ...

     test(“A test case”, () => {…})

     ...
  }
   
  describe("Another Inner describe", () => {
     ...

     test(“Another test case”, () => {…})

     ...
  }
})
```
__________________________________________________________________

#### `test(name: String, () =>, should_fail: bool)` - Defines an example/test case. You can use test() inside describe() blocks or independently.

Example: 

```
import { describe, test } from "matchstick-as/assembly/index"

describe(“Example test group”, () => {
    test(“Example test in a group”, () => {…}) // This case is expected to succeed.
    ...
})

test(“Independent test with should_fail set to true”, () => {…}, true) // This case is expected to fail. 


```
__________________________________________________________________

#### `beforeAll(() => {})` - Runs a code block before any of the tests in the file. If beforeAll is inside a describe block, it runs at the beginning of the describe block.

Examples: 

Code inside `beforeAll` will execute once before all describe blocks

```
import {describe, test, beforeAll } from "matchstick-as/assembly/index"

beforeAll(() => {
    …
})

describe(“Example test group”, () => {
    …
})

describe(“Another test group”, () => {
    …
})
```

Code inside `beforeAll` will execute once before all tests in the first describe block
```
import {describe, test, beforeAll } from "matchstick-as/assembly/index"


describe(“Example test group”, () => {
    beforeAll(() => {
        …
    })

    test(“Example test case”, () => {…})
    test(“Another test case”, () => {…})
})

describe(“Another test group”, () => {
    …
})
```
__________________________________________________________________

#### `afterAll(() => {})` - Runs a code block after all of the tests in the file. If afterAll is inside a describe block, it runs at the end of the describe block. 

Example:

Code inside `afterAll` will execute once after all describe blocks
```
import {describe, test, afterAll } from "matchstick-as/assembly/index"

afterAll(() => {
    …
})

describe(“Example test group”, () => {
    …
})

describe(“Another test group”, () => {
    …
})
```

Code inside `afterAll` will execute once after all tests in the first describe block
```
import {describe, test, afterAll } from "matchstick-as/assembly/index"


describe(“Example test group”, () => {
	afterAll(() => {
		…
	})

    test(“Example test case”, () => {…})
    test(“Another test case”, () => {…})
})

describe(“Another test group”, () => {
	…
})
```
__________________________________________________________________

#### `beforeEach(() => {})` - Runs a code block before every test. If beforeEach is inside a describe block, it runs before each test in that describe.  

Examples:
Code inside `beforeEach` will execute before each tests.
```
import {describe, test, beforeEach } from "matchstick-as/assembly/index"

beforeEach(() => {
    …
})

describe(“Example test group”, () => {
    …
})

describe(“Another test group”, () => {
    …
})
```

Code inside `beforeEach` will execute before each test in the first describe
```
import {describe, test, beforeEach } from "matchstick-as/assembly/index"


describe(“Example test group”, () => {
	beforeEach(() => {
		…
	})

    test(“Example test case”, () => {…})
    test(“Another test case”, () => {…})
})

describe(“Another test group”, () => {
	…
})
```
__________________________________________________________________

#### `afterEach(() => {})` - Runs a code block after every test. If afterEach is inside a describe block, it runs after each test in that describe.  

Examples:

Code inside `afterEach` will execute after every test.
```
import {describe, test, afterEach } from "matchstick-as/assembly/index"

afterEach(() => {
    …
})

describe(“Example test group”, () => {
    …
})

describe(“Another test group”, () => {
    …
})
```

Code inside `afterEach` will execute after each test in the first describe
```
import {describe, test, afterEach } from "matchstick-as/assembly/index"


describe(“Example test group”, () => {
    afterEach(() => {
        …
    })

    test(“Example test case”, () => {…})
    test(“Another test case”, () => {…})
})

describe(“Another test group”, () => {
	…
})
```

## <a href="#user-content-how-to-test" id="how-to-test">How To Test:</a>
There is a beta release if you want to try it out, but you'll need a couple of things before that:

1. `matchstick-as: 0.5.0.
4. You'll need to use matchstick v `0.5.0`. You can do that either by providing the `-v` flag to the `graph test` command `graph test -v 0.5.0`, or you can download the [binary](https://github.com/LimeChain/matchstick/releases/tag/0.5.0) for your OS, to your subgraph folder, change the permissions with  `chmod a+x` and then run the binary from the terminal, for example `./binary-macos-11`

That should be all!

If you encounter any bug or issues, or have any questions, feel free to leave a comment.

### CODE REVIEW

File: src/compiler/sources.rs [hunk 1]
```diff
@@ -145,31 +139,58 @@ fn are_imports_modified(in_file: &Path, wasm_modified: SystemTime) -> bool {
...
-    combined_path.push(format!("{}.ts", imported_file));
-    combined_path
-        .canonicalize()
-        .unwrap_or_else(|_| logging::critical!("{} does not exists!", &imported_file))
+    combined_path.push(imported_file);
+    if let Ok(abs_path) = combined_path.canonicalize() {
+        Ok(abs_path)
+    } else {
+        combined_path.set_extension("ts");
+        combined_path.canonicalize()
```
Comment (georg-getz) [inline src/compiler/sources.rs:156]: unwrap_or_else

Comment (dimitrovmaksim) [reply src/compiler/sources.rs:156]: Всъщносто обърках се малко, но в случая тук трябва да връща `Result` и се хендълва на няколко места със `let Ok(path) = get_import_absolute_path(...)`, така че всеки път, който не съществува, реално се игнорира.

File: src/context.rs [hunk 2]
```diff
@@ -178,7 +178,35 @@ impl<C: Blockchain> MatchstickInstanceContext<C> {
...
+    /// function _registerDescribe(name: string, funcIdx: u32): void
+    pub fn register_describe(
+        &mut self,
+        _gas: &GasCounter,
+        name: AscPtr<AscString>,
+        func_idx: u32,
+    ) -> Result<(), HostExportError> {
+        let name: String = asc_get(&self.wasm_ctx, name, &GasCounter::new())?;
+        self.meta_tests
+            .push((name, false, func_idx, String::from("describe")));
```
Comment (georg-getz) [inline src/context.rs:195]: 1 скоба в повече има

File: src/context.rs [hunk 3]
```diff
@@ -178,7 +178,35 @@ impl<C: Blockchain> MatchstickInstanceContext<C> {
...
+    /// function _registerHook(funcIdx: u32, role: string): void
+    pub fn register_hook(
+        &mut self,
+        _gas: &GasCounter,
+        func_idx: u32,
+        role: AscPtr<AscString>,
+    ) -> Result<(), HostExportError> {
+        let role: String = asc_get(&self.wasm_ctx, role, &GasCounter::new())?;
+        self.meta_tests
+            .push((String::from(""), false, func_idx, role));
```
Comment (georg-getz) [inline src/context.rs:209]: пак скоба

File: src/context.rs (hunk 2 above)
Comment (dimitrovmaksim) [reply src/context.rs:195]: 3 отварящи и 3 затварящи има, иначе щеше да се чупи

File: src/test_suite.rs [hunk 4]
```diff
@@ -122,33 +133,152 @@ impl<C: Blockchain> From<&MatchstickInstance<C>> for TestSuite {
...
-        for (name, should_fail, func_idx) in &matchstick
-            .instance_ctx
-            .borrow()
-            .as_ref()
+        let functions = matchstick.instance_ctx().meta_tests.clone();
+        build_test_group(matchstick, "", functions, &table)
+    }
+}
+
+// A recursive fucntion that builds the the test suite from a single test.ts file.
```
Comment (georg-getz) [inline src/test_suite.rs:141]: fucn

File: src/test_suite.rs [hunk 5]
```diff
@@ -122,33 +133,152 @@ impl<C: Blockchain> From<&MatchstickInstance<C>> for TestSuite {
...
+            Testable::Group(group) => {
+                update_test_hooks(group, before_each.clone(), after_each.clone());
+            }
+        }
     }
 }
+
+// In order to get all functions inside a describe() block, we need to create a clone
+// of the current MatchstickInstance, then fetch and execute the said describe() function by it's id
+// from the context of the cloned instance. This will trigger the reigstration of all the functions
```
Comment (georg-getz) [inline src/test_suite.rs:225]: reig

**Review by georg-getz — APPROVED** [review]

**Pull request merged**